Source/ACE.Server/Managers/ContractManager.cs

```diff
@@ -0,0 +1,283 @@
+        }
+
+        /// <summary>
+        /// Abandon a contract in the Player's registry
+        /// </summary>
+        public void Abandon(uint contractId)
+        {
+            Erase(contractId);
+        }
+
+        /// <summary>
+        /// Erase a contract in the Player's registry
+        /// </summary>
+        public void Erase(uint contractId)
+        {
+            if (Debug)
+                Console.WriteLine($"{Player.Name}.ContractManager.Erase({contractId})");
+
+            var contract = GetContract(contractId);
+
+            if (contract != null)
+            {
+                contract.DeleteContract = true;
+                Contracts.Remove(contract);
+                if (Player != null)
+                {
+                    Player.CharacterChangesDetected = true;
+                    Player.Session.Network.EnqueueSend(new GameEventSendClientContractTracker(Player.Session, contract));                    
+                }
+
+                RefreshMonitoredQuestFlags();
+            }
+        }
+
+        /// <summary>
+        /// Erase all contracts in registry
+        /// </summary>
+        public void EraseAll()
+        {
+            if (Debug)
+                Console.WriteLine($"{Player.Name}.ContractManager.EraseAll");
+
+            var contracts = Contracts.ToList();
+            foreach (var contract in contracts)
+            {
+                contract.DeleteContract = true;
+                Contracts.Remove(contract);
+                if (Player != null)
+                {
+                    Player.CharacterChangesDetected = true;
+                    Player.Session.Network.EnqueueSend(new GameEventSendClientContractTracker(Player.Session, contract));                    
+                }
+            }
+
+            RefreshMonitoredQuestFlags();
+        }
+
+        public void NotifyOfQuestUpdate(string questName)
+        {
+            foreach (var contracts in MonitoredQuestFlags)
+            {
+                if (contracts.Value.Contains(questName.ToLower()))
+                    Update(contracts.Key);
+            }
+        }
+
+        private void Update(uint contractId)
+        {
+            if (Debug)
+                Console.WriteLine($"{Player.Name}.ContractManager.Update");
+
+            if (Player != null)
+            {
+                Player.Session.Network.EnqueueSend(new GameEventSendClientContractTracker(Player.Session, contractId));
+            }
+        }
+    }
+
+    public static class ContractManagerExtensions
+    {
+        public static void Write(this BinaryWriter writer, ContractManager contractManager)
+        {
+            writer.Write(contractManager.ContractTrackerTable);
+        }
+
+        public static void Write(this BinaryWriter writer, Dictionary<uint, ContractTracker> contractTrackerHash)
+        {
+            PackableHashTable.WriteHeader(writer, contractTrackerHash.Count);
+            foreach (var kvp in contractTrackerHash)
```

Comment:
this table needs to be sorted by key % 32

(confirmed it seems to always use 32 tablesize in the pcaps)